src/Chapter4.hs

```diff
@@ -470,12 +484,16 @@ Applicatives can be found in many applications:
 
 Implement the Applicative instance for our 'Secret' data type from before.
 -}
+
+
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
-
+    pure = Reward
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap x) (Trap _) = Trap x
+    (<*>) (Reward _) (Trap x) = Trap x
+    (<*>) (Trap x) (Reward _) = Trap x
+    (<*>) (Reward f) (Reward x) = Reward (f x)
```

Comment:
You can notice that cases with the `Reward` constructor on the left side are very similar to the `Functor` instance for `Secret`, so you can reuse the `fmap` here to have a slightly shorter implementation :slightly_smiling_face:

Reply:
Oh right, I think it is often the case that Functor's `fmap` can be sometimes reused for the implementations of the `Applicative`.